ArmPkg/Drivers/CpuDxe/PageTableMemoryAllocation.c

```diff
@@ -0,0 +1,202 @@
+/** @file PageTableAlloc.c
+
+Logic facilitating the pre-allocation of page table memory.
+
+Copyright (C) Microsoft Corporation. All rights reserved.
+SPDX-License-Identifier: BSD-2-Clause-Patent
+**/
+
+#include <Uefi.h>
+#include <Library/BaseLib.h>
+#include <Library/DebugLib.h>
+#include <Library/MemoryAllocationLib.h>
+#include <Library/UefiBootServicesTableLib.h>
+#include <Protocol/ArmPageTableMemoryAllocation.h>
+
+#pragma pack(1)
+
+typedef struct {
+  UINT32        Signature;
+  UINTN         Offset;
+  UINTN         FreePages;
+  LIST_ENTRY    NextPool;
+} PAGE_TABLE_POOL;
+
+#pragma pack()
+
+#define MIN_PAGES_AVAILABLE        5
+#define PAGE_TABLE_POOL_SIGNATURE  SIGNATURE_32 ('P','T','P','L')
+
+UINTN       TotalFreePages     = 0;
+BOOLEAN     mPageTablePoolLock = FALSE;
+LIST_ENTRY  mPageTablePoolList = INITIALIZE_LIST_HEAD_VARIABLE (mPageTablePoolList);
+
+/**
+  Allocates pages for page table memory and adds them to the pool list.
+
+  @param[in] PoolPages The number of pages to allocate. The actual number of pages
+                       reserved will be at least 512 - the number of pages required to describe a
+                       fully split 2MB region.
+
+  @retval EFI_SUCCESS             More pages successfully added to the pool list
+  @retval EFI_INVALID_PARAMETER   PoolPages is zero
+  @retval EFI_ABORTED             Software lock is held by another call to this function
+  @retval EFI_OUT_OF_RESOURCES    Unable to allocate pages
+**/
+EFI_STATUS
+GetMorePages (
+  IN  UINTN  PoolPages
+  )
+{
+  PAGE_TABLE_POOL  *Buffer;
+
+  if (PoolPages == 0) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  PoolPages++;
+  PoolPages = ALIGN_VALUE (PoolPages, EFI_SIZE_TO_PAGES (SIZE_2MB)); // Add one page for the header
+  Buffer    = (PAGE_TABLE_POOL *)AllocateAlignedPages (PoolPages, BASE_2MB);

```

Comment:
nit: why is this BASE_2MB instead of SIZE_2MB?

Reply:
I must have searched Base.h for a 2MB size macro two separate times and found the two results. Fixed